### PR TITLE
ICS31 integration tests

### PR DESCRIPTION
**Pull request by ljoss17**
Closes: #2967

## Description

This PR adds integration test for ICS31 Cross Chain Queries, using Stride <-> Gaia chains.

______

### PR author checklist:
- [x] Added changelog entry, using [`unclog`](https://github.com/informalsystems/unclog).
- [x] Added tests: integration (for Hermes) or unit/mock tests (for modules).
- [x] Linked to GitHub issue.
- [x] Updated code comments and documentation (e.g., `docs/`).
- [x] Tagged *one* reviewer who will be the one responsible for shepherding this PR.

### Reviewer checklist:

- [ ] Reviewed `Files changed` in the GitHub PR explorer.
- [ ] Manually tested (in case integration/unit/mock tests are absent).

### CODE REVIEW

File: .github/workflows/integration.yaml [hunk 1]
```diff
@@ -261,6 +261,46 @@ jobs:
             test -p ibc-integration-test --features forward-packet --no-fail-fast -- \
             --nocapture --test-threads=1 forward::
 
+  ics31:
+    runs-on: ubuntu-20.04
+    strategy:
+      fail-fast: false
+      matrix:
+        chain:
```
Comment (soareschen) [inline .github/workflows/integration.yaml:269]: I think you need to add the new `stride-no-admin` package here.

Comment (ljoss17) [reply .github/workflows/integration.yaml:269]: Testing requires to have one Stride chain and one Gaia chain, if the `stride-no-admin` is added to the `matrix` wouldn't result in having one test running 2 Gaia chains and another test running two Stride chains ?
The idea here was to have the possibility to test multiple Gaia versions if we want/need to, by simply adding them to the `matrix`. Resulting in every test being:

* `Chain from matrix` <-> `Stride chain`

What do you think ?

Comment (soareschen) [reply .github/workflows/integration.yaml:269]: Hmm, I think if you want a CI matrix, then putting stride in it also result in all the combinations with stride executed.

File: tools/integration-test/src/tests/ics31.rs [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        chains.node_b.chain_driver().assert_eventual_wallet_amount(
+            &wallet_b.address(),
+            &denom_b.with_amount(a_to_b_amount).as_ref(),
+        )?;
+
+        info!("Registering host-zone");
+
+        simple_exec(
+            "stride",
+            "strided",
```
Comment (soareschen) [inline tools/integration-test/src/tests/ics31.rs:117]: Does this command always have to be executed with `stride`? Might be better to use the chain driver's command path to execute the command.

File: tools/test-framework/src/chain/cli/query.rs [hunk 3]
```diff
@@ -96,3 +96,60 @@ fn yaml_to_json_value(value: yaml::Value) -> Result<json::Value, Error> {
 
     Ok(parsed)
 }
+
+pub fn query_cross_chain_query_list(
```
Comment (soareschen) [inline tools/test-framework/src/chain/cli/query.rs:100]: This looks like two functionalities mixed into one function. There should be one query function to return the result from `simple_exec`, and another assert function that calls the function, and asserts that the result is as expected.

File: tools/test-framework/src/chain/driver.rs [hunk 4]
```diff
@@ -188,6 +198,15 @@ impl ChainDriver {
         )
     }
 
+    pub fn query_cross_chain_query_list(&self, pending_queries: bool) -> Result<(), Error> {
```
Comment (soareschen) [inline tools/test-framework/src/chain/driver.rs:201]: May be better to introduce this as an extension method in `src/chain/ext`.

File: tools/test-framework/src/chain/driver.rs [hunk 5]
```diff
@@ -219,4 +238,31 @@ impl ChainDriver {
 
         Ok(())
     }
+
+    /**
+       Depending on the value of `pending_queries`, assert that a
+       cross chain is pending or that no cross chain queries are
+       pending..
+    */
+    pub fn assert_eventual_cross_chain_query_relayed(
```
Comment (soareschen) [inline tools/test-framework/src/chain/driver.rs:247]: May be better to introduce this as an extension method in `src/chain/ext`.

File: tools/test-framework/src/chain/tagged.rs [hunk 6]
```diff
@@ -62,6 +62,9 @@ pub trait TaggedChainDriverExt<Chain> {
         token: &TaggedTokenRef<Chain>,
     ) -> Result<(), Error>;
 
+    fn assert_eventual_cross_chain_query_relayed(&self, pending_queries: bool)
```
Comment (soareschen) [inline tools/test-framework/src/chain/tagged.rs:65]: May be better to introduce this as an extension method in `src/chain/ext`.

**Review by soareschen — CHANGES_REQUESTED** [review]

File: tools/integration-test/src/tests/ics31.rs (hunk 2 above)
Comment (ljoss17) [reply tools/integration-test/src/tests/ics31.rs:117]: `gaiad` doesnt have the CLI `<command_path> tx stakeibc register-host-zone`.
I could create move this to a new subdirectory in the `test-framework`:

`test-framework/src/chain/cli/chain-specific/stride/host-zone.rs`

What do you think ?

Comment (soareschen) [reply tools/integration-test/src/tests/ics31.rs:118]: The test itself should not care whether the chain executable supports a given command. If the command is not supported, the test can just fail. We only need to configure the CI matrix such that the supported chain executables are given in the `CHAIN_COMMAND_PATH`.

**Review by soareschen — APPROVED** [review]

**Pull request merged**